duckduckgo/iOS
# Import / Export Bookmarks

Task/Issue URL: https://app.asana.com/0/72649045549333/1201780742295295/f
Tech Design URL:
CC: @brindy 

**Description**:

New functionality / enhancement that allows the importing and exporting of bookmarks to / from html files

Additionally I've made a few modifications to BookmarksCoreDataStorage + added tests to support this new functionality to allow for:
 * MockBookmarksCoreDataStorage so that existing functionality and new functionality to be unit tested
 * swift concurrency (async await)

**Steps to test this PR**:

**Test Scenario 1**

1. Start with a previous build of the app, with favorites, bookmarks & bookmark folders
2. Install this PR
3. Export bookmarks 
4. Confirm appropriate export success toast message is shown to user 
a. Save to Files = "Your bookmarks have been exported." 
b. Shared to any other app = "Your bookmarks have been shared."
5. Delete + reinstall the app / delete all the bookmarks & favorites in the app
6. Import the bookmarks file from step 3
7. All previous favorites & bookmarks should be re-added
8. Confirm import success toast message is shown to user

**Test Scenario 2**
1. Ensuring the app has a mix of favorites, top level bookmarks & bookmark folders, export bookmarks to file
2. Confirm export success toast message is shown to user
3. Import the bookmarks file
4. Confirm import success toast message is shown to user
5. Top level favorites and bookmarks should not be duplicated, only bookmark folders

**Test Scenario 3**
1. Export bookmarks from another browser e.g. Chrome / Safari
2. Import the bookmarks into the app
3. Confirm import success toast message is shown to user
4. Bookmarks and folders should import correctly

**Test Scenario 4**
1. Export bookmarks from the app
2. Confirm export success toast message is shown to user
3. Import the bookmarks into another browser e.g. Chrome or Safari and confirm bookmarks are imported

**Test Scenario 5**
1. Try to import an invalid html bookmark file (there is one in udner Test -> Mock files)
2. Confirm export failed toast message is shown to user
	
**Test Scenario 6** 
1. Navigate to a bookmark subfolder.
2. Confirm More button is disabled

**Test Scenario 7** 
1. Delete all bookmarks, favorites and folders
2. Confirm the More -> Export button is disabled

**Test Scenario 8** 
1. Add items to Safari reading list and export bookmarks 
2. Import the Safari bookmarks html and confirm the reading list is not imported

**Test Scenario 9** 
1. Import bookmarks into the app
2. Tap on an imported bookmark that does not have a favicon to load that bookmark in browser
3. Go back to the list of bookmarks and confirm the bookmark now has a favicon   

**Test Scenario 10** 
1. Using a device running iOS 13, confirm no import or export functionality is available i.e. More button is not in the toolbar and the import footer button is not visible2. 

**Unit Tests**
- Run Unit Tests for `BookmarkCoreDataStorageTests`, `BookmarksImporterTests` & `BookmarksExporterTests`  

**Pixels**

Test the following pixels fire appropriately

- case bookmarkImportSuccess = "m_bi_s"
- case bookmarkImportFailure = "m_bi_e"
- case bookmarkExportSuccess = "m_be_a"
- case bookmarkExportFailure = "m_be_e"




**Copy Testing**:


**Orientation Testing**:


**Device Testing**:


**OS Testing**:


**Theme Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**
